### ISSUE TITLE
Output column "wt" could contain a single AA, not the whole WT sequence?

### ISSUE THREAD

**Issue by mgalardini** [allydunham/sequence_unet#4]
Sorry for the multiple posts! This is not a crash but rather a suggestion for a small improvement:

`python make_preds.py --fasta acrB.faa -m . freq_classifier > acrB.tsv`

These are the first 2 rows:

```
gene    position        wt      mut     pred
sp|P31224|ACRB_ECOLI    1       MPNFFIDRPIFAWVIAIIIMLAGGLAILKLPVAQYPTIAPPAVTISASYPGADAKTVQDTVTQVIEQNMNGIDNLMYMSSNSDSTGTVQITLTFESGTDADIAQVQVQNKLQLAMPLLPQEVQQQGVSVEKSSSSFLMVVGVINTDGTMTQEDISDYVAANMKDAISRTSGVGDVQLFGSQYAMRIWMNPNELNKFQLTPVDVITAIKAQNAQVAAGQLGGTPPVKGQQLNASIIAQTRLTSTEEFGKILLKVNQDGSRVLLRDVAKIELGGENYDIIAEFNGQPASGLGIKLATGANALDTAAAIRAELAKMEPFFPSGLKIVYPYDTTPFVKISIHEVVKTLVEAIILVFLVMYLFLQNFRATLIPTIAVPVVLLGTFAVLAAFGFSINTLTMFGMVLAIGLLVDDAIVVVENVERVMAEEGLPPKEATRKSMGQIQGALVGIAMVLSAVFVPMAFFGGSTGAIYRQFSITIVSAMALSVLVALILTPALCATMLKPIAKGDHGEGKKGFFGWFNRMFEKSTHHYTDSVGGILRSTGRYLVLYLIIVVGMAYLFVRLPSSFLPDEDQGVFMTMVQLPAGATQERTQKVLNEVTHYYLTKEKNNVESVFAVNGFGFAGRGQNTGIAFVSLKDWADRPGEENKVEAITMRATRAFSQIKDAMVFAFNLPAIVELGTATGFDFELIDQAGLGHEKLTQARNQLLAEAAKHPDMLTSVRPNGLEDTPQFKIDIDQEKAQALGVSINDINTTLGAAWGGSYVNDFIDRGRVKKVYVMSEAKYRMLPDDIGDWYVRAADGQMVPFSAFSSSRWEYGSPRLERYNGLPSMEILGQAAPGKSTGEAMELMEQLASKLPTGVGYDWTGMSYQERLSGNQAPSLYAISLIVVFLCLAALYESWSIPFSVMLVVPLGVIGALLAATFRGLTNDVYFQVGLLTTIGLSAKNAILIVEFAKDLMDKEGKGLIEATLDAVRMRLRPILMTSLAFILGVMPLVISTGAGSGAQNAVGTGVMGGMVTATVLAIFFVPVFFVVVRRRFSRKNEDIEHSHTVDHH      A       0.9902427
sp|P31224|ACRB_ECOLI    1       MPNFFIDRPIFAWVIAIIIMLAGGLAILKLPVAQYPTIAPPAVTISASYPGADAKTVQDTVTQVIEQNMNGIDNLMYMSSNSDSTGTVQITLTFESGTDADIAQVQVQNKLQLAMPLLPQEVQQQGVSVEKSSSSFLMVVGVINTDGTMTQEDISDYVAANMKDAISRTSGVGDVQLFGSQYAMRIWMNPNELNKFQLTPVDVITAIKAQNAQVAAGQLGGTPPVKGQQLNASIIAQTRLTSTEEFGKILLKVNQDGSRVLLRDVAKIELGGENYDIIAEFNGQPASGLGIKLATGANALDTAAAIRAELAKMEPFFPSGLKIVYPYDTTPFVKISIHEVVKTLVEAIILVFLVMYLFLQNFRATLIPTIAVPVVLLGTFAVLAAFGFSINTLTMFGMVLAIGLLVDDAIVVVENVERVMAEEGLPPKEATRKSMGQIQGALVGIAMVLSAVFVPMAFFGGSTGAIYRQFSITIVSAMALSVLVALILTPALCATMLKPIAKGDHGEGKKGFFGWFNRMFEKSTHHYTDSVGGILRSTGRYLVLYLIIVVGMAYLFVRLPSSFLPDEDQGVFMTMVQLPAGATQERTQKVLNEVTHYYLTKEKNNVESVFAVNGFGFAGRGQNTGIAFVSLKDWADRPGEENKVEAITMRATRAFSQIKDAMVFAFNLPAIVELGTATGFDFELIDQAGLGHEKLTQARNQLLAEAAKHPDMLTSVRPNGLEDTPQFKIDIDQEKAQALGVSINDINTTLGAAWGGSYVNDFIDRGRVKKVYVMSEAKYRMLPDDIGDWYVRAADGQMVPFSAFSSSRWEYGSPRLERYNGLPSMEILGQAAPGKSTGEAMELMEQLASKLPTGVGYDWTGMSYQERLSGNQAPSLYAISLIVVFLCLAALYESWSIPFSVMLVVPLGVIGALLAATFRGLTNDVYFQVGLLTTIGLSAKNAILIVEFAKDLMDKEGKGLIEATLDAVRMRLRPILMTSLAFILGVMPLVISTGAGSGAQNAVGTGVMGGMVTATVLAIFFVPVFFVVVRRRFSRKNEDIEHSHTVDHH      C       0.99634844
```

I would have expected the `wt` column to report `M` and not the whole sequence. Seems like a potential waste of disk space. I looked a bit at the code and I think [maybe something should be changed around here](https://github.com/allydunham/sequence_unet/blob/master/sequence_unet/predict.py#L250)?

Thanks again for this amazing piece of software!

**Comment by allydunham** [comment]
No worries, thanks for testing it out throughly! I'm away for a week so I'll look properly when I get home.

This does seem like an oversight/odd behaviour though. Interestingly when I tested the script for your other issue (on another machine) I didn't notice the output looking like this though and neither do the output files on this machine, so perhaps it actually is a bug.

I wonder if it's a change in behaviour with pandas/numpy on some version? In my output it looks like the WT sequence is getting assigned to the column as a vector whereas for you it seems to be being replicated in each column row. What package versions are you running with?

**Comment by mgalardini** [comment]
Ah, maybe that could be it, pandas has changed a lot lately.

Here's my conda environment:
```
# packages in environment at /fast-storage/miniconda3/envs/sequence_unet:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    conda-forge
_openmp_mutex             5.1                       1_gnu  
_tflow_select             2.3.0                       mkl  
abseil-cpp                20211102.0           h27087fc_1    conda-forge
absl-py                   1.4.0              pyhd8ed1ab_0    conda-forge
aiohttp                   3.8.1            py39hb9d737c_1    conda-forge
aiosignal                 1.3.1              pyhd8ed1ab_0    conda-forge
anndata                   0.8.0                    pypi_0    pypi
astunparse                1.6.3              pyhd8ed1ab_0    conda-forge
async-timeout             4.0.3              pyhd8ed1ab_0    conda-forge
attrs                     22.2.0                   pypi_0    pypi
biopython                 1.79                     pypi_0    pypi
biotite                   0.35.0                   pypi_0    pypi
blas                      1.0                         mkl    conda-forge
bleach                    5.0.1                    pypi_0    pypi
blinker                   1.6.2              pyhd8ed1ab_0    conda-forge
brotlipy                  0.7.0           py39hb9d737c_1004    conda-forge
build                     0.9.0                    pypi_0    pypi
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2023.01.10           h06a4308_0  
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cachetools                5.3.1              pyhd8ed1ab_0    conda-forge
certifi                   2022.12.7        py39h06a4308_0  
cffi                      1.15.1           py39h74dc2b5_0  
charset-normalizer        2.1.1              pyhd8ed1ab_0    conda-forge
click                     8.1.7           unix_pyh707e725_0    conda-forge
clize                     5.0.0                    pypi_0    pypi
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
commonmark                0.9.1                    pypi_0    pypi
contourpy                 1.0.6                    pypi_0    pypi
cryptography              3.4.8            py39h95dcef6_1    conda-forge
cudatoolkit               11.3.1               h2bc3f7f_2  
cycler                    0.11.0                   pypi_0    pypi
docutils                  0.19                     pypi_0    pypi
evolocity                 1.0                      pypi_0    pypi
exceptiongroup            1.1.0                    pypi_0    pypi
fair-esm                  2.0.0                    pypi_0    pypi
flatbuffers               2.0.6                h27087fc_0    conda-forge
fonttools                 4.38.0                   pypi_0    pypi
frozenlist                1.3.0            py39hb9d737c_1    conda-forge
gast                      0.4.0              pyh9f0ad1d_0    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
google-auth               2.22.0             pyh1a96a4e_0    conda-forge
google-auth-oauthlib      0.4.6              pyhd8ed1ab_0    conda-forge
google-pasta              0.2.0              pyh8c360ce_0    conda-forge
grpc-cpp                  1.46.1               h33aed49_1  
grpcio                    1.38.1           py39hff7568b_0    conda-forge
h5py                      3.7.0                    pypi_0    pypi
hdf5                      1.12.1          nompi_h2750804_100    conda-forge
hhsuite                   3.3.0           py39pl5321h67e14b5_5    bioconda
hmmer                     3.3.2                h87f3376_2    bioconda
icu                       58.2              hf484d3e_1000    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
igraph                    0.10.2                   pypi_0    pypi
importlib-metadata        5.2.0                    pypi_0    pypi
iniconfig                 1.1.1                    pypi_0    pypi
intel-openmp              2022.1.0          h9e868ea_3769  
jaraco-classes            3.2.3                    pypi_0    pypi
jeepney                   0.8.0                    pypi_0    pypi
jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge
joblib                    1.2.0              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h166bdaf_1    conda-forge
keras                     2.10.0             pyhd8ed1ab_0    conda-forge
keyring                   23.13.1                  pypi_0    pypi
kiwisolver                1.4.4                    pypi_0    pypi
krb5                      1.19.4               h568e23c_0  
ld_impl_linux-64          2.38                 h1181459_1  
libblas                   3.9.0            16_linux64_mkl    conda-forge
libcblas                  3.9.0            16_linux64_mkl    conda-forge
libcurl                   8.1.1                h91b91d3_0  
libedit                   3.1.20221030         h5eee18b_0  
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.3                  he6710b0_2  
libgcc-ng                 11.2.0               h1234567_1  
libgfortran-ng            12.2.0              h69a702a_19    conda-forge
libgfortran5              12.2.0              h337968e_19    conda-forge
libgomp                   11.2.0               h1234567_1  
liblapack                 3.9.0            16_linux64_mkl    conda-forge
libnghttp2                1.46.0               hce63b2e_0  
libnsl                    2.0.0                h7f98852_0    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libprotobuf               3.20.3               he621ea3_0  
libssh2                   1.10.0               ha56f1ee_2    conda-forge
libstdcxx-ng              11.2.0               h1234567_1  
llvmlite                  0.39.1                   pypi_0    pypi
louvain                   0.8.0                    pypi_0    pypi
mafft                     7.310                h87f3376_5    bioconda
markdown                  3.4.4              pyhd8ed1ab_0    conda-forge
markupsafe                2.1.1            py39hb9d737c_1    conda-forge
matplotlib                3.6.2                    pypi_0    pypi
mkl                       2022.1.0           hc2b9512_224  
more-itertools            9.0.0                    pypi_0    pypi
msgpack                   1.0.4                    pypi_0    pypi
multidict                 6.0.2            py39hb9d737c_1    conda-forge
natsort                   8.2.0                    pypi_0    pypi
ncurses                   6.3                  h5eee18b_3  
networkx                  2.8.8                    pypi_0    pypi
numba                     0.56.4                   pypi_0    pypi
numpy                     1.23.4                   pypi_0    pypi
oauthlib                  3.2.2              pyhd8ed1ab_0    conda-forge
od                        2.0.2                    pypi_0    pypi
openssl                   1.1.1t               h7f8727e_0  
opt_einsum                3.3.0              pyhd8ed1ab_1    conda-forge
packaging                 21.3                     pypi_0    pypi
pandas                    1.5.1                    pypi_0    pypi
patsy                     0.5.3                    pypi_0    pypi
pep517                    0.13.0                   pypi_0    pypi
perl                      5.32.1          2_h7f98852_perl5    conda-forge
pillow                    9.3.0                    pypi_0    pypi
pip                       22.2.2           py39h06a4308_0  
pkginfo                   1.9.2                    pypi_0    pypi
pluggy                    1.0.0                    pypi_0    pypi
proteinnetpy              1.0.1                    pypi_0    pypi
protobuf                  3.20.3           py39h6a678d5_0  
pyasn1                    0.4.8                      py_0    conda-forge
pyasn1-modules            0.2.7                      py_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyg                       2.1.0           py39_torch_1.12.0_cu113    pyg
pygments                  2.13.0                   pypi_0    pypi
pyjwt                     2.8.0              pyhd8ed1ab_0    conda-forge
pynndescent               0.5.8                    pypi_0    pypi
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
pytest                    7.2.0                    pypi_0    pypi
python                    3.9.13               haa1d7c7_2  
python-dateutil           2.8.2                    pypi_0    pypi
python-flatbuffers        23.5.26            pyhd8ed1ab_0    conda-forge
python-igraph             0.10.2                   pypi_0    pypi
python_abi                3.9                      2_cp39    conda-forge
pytorch                   1.12.1          py3.9_cuda11.3_cudnn8.3.2_0    pytorch
pytorch-cluster           1.6.0           py39_torch_1.12.0_cu113    pyg
pytorch-mutex             1.0                        cuda    pytorch
pytorch-scatter           2.0.9           py39_torch_1.12.0_cu113    pyg
pytorch-sparse            0.6.15          py39_torch_1.12.0_cu113    pyg
pytz                      2022.6                   pypi_0    pypi
pyu2f                     0.1.5              pyhd8ed1ab_0    conda-forge
re2                       2022.04.01           h27087fc_0    conda-forge
readline                  8.2                  h5eee18b_0  
readme-renderer           37.3                     pypi_0    pypi
requests                  2.28.1             pyhd8ed1ab_1    conda-forge
requests-oauthlib         1.3.1              pyhd8ed1ab_0    conda-forge
requests-toolbelt         0.10.1                   pypi_0    pypi
rfc3986                   2.0.0                    pypi_0    pypi
rich                      12.6.0                   pypi_0    pypi
rsa                       4.9                pyhd8ed1ab_0    conda-forge
scanpy                    1.9.1                    pypi_0    pypi
scikit-learn              1.0.2            py39h4dfa638_0    conda-forge
scipy                     1.9.3                    pypi_0    pypi
seaborn                   0.12.1                   pypi_0    pypi
secretstorage             3.3.3                    pypi_0    pypi
session-info              1.0.0                    pypi_0    pypi
setuptools                65.5.0           py39h06a4308_0  
sigtools                  4.0.1                    pypi_0    pypi
six                       1.16.0             pyh6c4a22f_0    conda-forge
snappy                    1.1.9                hbd366e4_1    conda-forge
sqlite                    3.39.3               h5082296_0  
statsmodels               0.13.5                   pypi_0    pypi
stdlib-list               0.8.0                    pypi_0    pypi
tensorboard               2.10.1             pyhd8ed1ab_0    conda-forge
tensorboard-data-server   0.6.0            py39hd97740a_2    conda-forge
tensorboard-plugin-wit    1.8.1              pyhd8ed1ab_0    conda-forge
tensorflow                2.10.0          mkl_py39hdc5b6f4_0  
tensorflow-base           2.10.0          mkl_py39hb9daa73_0  
tensorflow-estimator      2.10.0           py39h06a4308_0  
termcolor                 2.3.0              pyhd8ed1ab_0    conda-forge
texttable                 1.6.4                    pypi_0    pypi
threadpoolctl             3.1.0              pyh8a188c0_0    conda-forge
tk                        8.6.12               h1ccaba5_0  
tomli                     2.0.1                    pypi_0    pypi
tqdm                      4.64.1             pyhd8ed1ab_0    conda-forge
twine                     4.0.2                    pypi_0    pypi
typing-extensions         4.3.0            py39h06a4308_0  
typing_extensions         4.3.0            py39h06a4308_0  
tzdata                    2022f                h04d1e81_0  
umap-learn                0.5.3                    pypi_0    pypi
urllib3                   1.26.12                  pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
werkzeug                  2.3.7              pyhd8ed1ab_0    conda-forge
wheel                     0.37.1             pyhd3eb1b0_0  
wrapt                     1.14.1           py39hb9d737c_0    conda-forge
xz                        5.2.6                h5eee18b_0  
yarl                      1.7.2            py39hb9d737c_2    conda-forge
zipp                      3.11.0                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_0  
```

tl;dr I am running pandas 1.5.1

**Comment by allydunham** [comment]
I'm on numpy 1.21.5 and pandas 1.3.5 on this machine (probably similar on my work machine), so that may be it. I don't use that much pandas so hopefully there's a setup that works across versions, otherwise I might need to pin the required version which would be a bit annoying.

I don't use pandas very much in general, so if you have an idea for how to get that line to always vectorise across the whole column that would be great!

**Comment by mgalardini** [comment]
I tried this in [this line](https://github.com/allydunham/sequence_unet/blob/master/sequence_unet/predict.py#L250):

```
df['wt'] = np.array(seq.seq)
```

and it worked with pandas 1.5.1. I'm guessing something has changed in the way strings are digested with the square bracket operator. Forcing it to an array first apparently solves the issue.

**Comment by allydunham** [comment]
I've updated the predict_sequence function more generally, so it should now correctly work with Biopython SeqRecord iterators and simple iterators of sequence strings (or string-like things), which was always the intention. It now converts all input to strings and then uses `df['wt'] = [i for i in seq.seq]`, which works accross Pandas versions I think. The SeqRecord interaction with pandas seemed a bit unpredictable, with some versions acting like an array and others a string. This way should hopefully be consistent going forward as it just uses base python stuff.
